lib/asciidoctor/converter/semantic_html5.rb

```diff
@@ -96,6 +96,81 @@ def convert_thematic_break node
     '<hr>'
   end
 
+  def convert_ulist node
+    attributes = common_html_attributes node.id, node.role
+    ret = [%(<ul#{attributes}>)]
+    ret << %(<strong class="title">#{node.title}</strong>) if node.title?
+    node.items.each do |item|
+    sub_content = if item.blocks?
+      %(
+#{item.content}
+)
+    else
+      ""
+    end
+    ret << %(<li>#{item.text}#{sub_content}</li>)
```

Comment:
Do we need a wrapping element around `item.text`?
Currently, we have:

```html
<ul>
<li>alice
<ul>
<li>subpoint 1</li>
<li>subpoint 2</li>
</ul>
</li>
</ul>
```

With the following structure, the text "alice" cannot be easily styled. So, maybe we should do:

```html
<ul>
<li><span class="listitem">alice</span>
<ul>
<li>subpoint 1</li>
<li>subpoint 2</li>
</ul>
</li>
</ul>
```

I believe it's unnecessary to add a class name on the nested `<ul>` since you can use the CSS selector `ul ul` to target nested lists.

Reply:
> Do we need a wrapping element around item.text?

Yes. It's absolutely essential. Otherwise, there is no way to target it with CSS, and that only gets us back into the dilemma we have with the current HTML of not being designer friendly.

Reply:
> I believe it's unnecessary to add a class name on the nested <ul> since you can use the CSS selector ul ul to target nested lists.

We need at least one role on any top-level element the converter creates to make it clear that the converter made that element. Otherwise, we can't make a distinction between lists created from an AsciiDoc list or a list created from some other source. It's not enough to assume that CSS can target it.

Reply:
> `<span class="listitem">alice</span>`

Since the whole element is already a list item, I prefer describing the part of it. That's why I like "principle" like in the EPUB 3 converter. I'd even consider "text" since that's what we call it in the Asciidoctor DOM.

Reply:
> Otherwise, we can't make a distinction between lists created from an AsciiDoc list or a list created from some other source. It's not enough to assume that CSS can target it.

So `<ul class="list">` and `<span class="text">alice</span>`?
`principle` might be more semantic but less common (and difficult to remember).

Speaking of naming, I wonder if we should rename `menuitem` to `item` (on the `menu` macro) since the element is already in a "menu" element? Otherwise, someone could argue that we should have use `listtext` (instead of `text`) to be consistent.

Reply:
> So `<ul class="list">` and `<span class="text">alice</span>`?

I would prefer matching the AsciiDoc block context names, in which case it would be `ulist` on the `<ul>` tag.

There's one other way to look at it. We could say that `<ul>` already tells us the context in this case. Instead, we add the common/generic identifier `block`. The reason for this is to be able to style all block elements consistently, like we see here: https://gitlab.com/antora/antora-ui-default/-/blob/master/src/css/doc.css#L200-217 As you can see, having different names without a common class is annoying. So perhaps:

```
<ul class="block">
```

That would also mean for paragraphs, we use:

```
<p class="block">
```

> I wonder if we should rename menuitem to item (on the menu macro) since the element is already in a "menu" element?

That's a good point. There's really no right or wrong way here, just what feels most comfortable. What's most important is that we're consistent about it.

Reply:
> There's one other way to look at it. We could say that `<ul>` already tells us the context in this case. Instead, we add the common/generic identifier block

I see, that's a good idea but what about `<figure>` used for image and source blocks? In this case, we will need to make a distinction using the block name.
As an aside, and speaking of source blocks, we are currently using `listing` even if it's a source block. I think we should use `source`. 

> As you can see, having different names without a common class is annoying

Why not...

```html
<ul class="block"></ul>
<ol class="block"></ol>
<dl class="block"></dl>
<dl class="qanda block"></dl>
<dl class="horizontal block"></dl>
<figure class="source block"></figure>
<figure class="image block"></figure>
<figure class="listing block"></figure>
```

The thing is `block` is a common name and if we really want to make a clear distinction between elements created from Asciidoctor and elements created from some other source then we should probably use a more unique name: `ablock`, `adocblock`, `asciidoctorblock`.
We could do the same for inline elements: `ainline`, `adocinline`, `asciidoctorinline`?

Reply:
On a related note, we could also add a common/generic class on headings? https://gitlab.com/antora/antora-ui-default/-/blob/master/src/css/doc.css#L21-26

It's less annoying to declare, because h1, h2, h3... is known, but if you want to make sure that you are only apply a style on headings created by Asciidoctor then it might be useful.

Reply:
> As an aside, and speaking of source blocks, we are currently using listing even if it's a source block. I think we should use source.

That seems reasonable. The distinction in AsciiDoc has always been fuzzy, so I'm open to exploring making it clearer. (I'd even be happy with "code" instead of "source" like in Asciidoctor PDF).

> The thing is block is a common name

I think we need to be careful not to go overboard. The idea is that we are trying to be as concise as possible without making it hard to style. I think a generic name is the right approach. Once we start adding "AsciiDoc" as a prefix, it's just going to get ugly. We can explore other generic names, but right now "block" seems to be comfortable to me.

Reply:
> On a related note, we could also add a common/generic class on headings?

I'd be okay with that. I like just "heading" as a name here.

Btw, when in doubt, we should look at how some CSS frameworks are naming classes. That might lead us in the right direction when we get lost.

Reply:
How about rather than `block` or `adocblock` as the catchall class, just `adoc`, concise and specific. to take the example from above

```
<ul class="adoc"></ul>
<ol class="adoc"></ol>
<dl class="adoc"></dl>
<dl class="qanda adoc"></dl>
<dl class="horizontal adoc"></dl>
<figure class="source adoc"></figure>
<figure class="image adoc"></figure>
<figure class="listing adoc"></figure>
```

Reply:
That's a very strong suggestion. I'm going to give that some consideration.

The only drawback is that it's not clear that it's a block rather than an inline element. And block elements almost always need to share styles.

We could consider "ablock" and "ainline" (or "aspan"). But even as I type that, remember we still have a root element we can use to scope styles like we do in the stylesheet for the default UI. That's where we could use the name "asciidoc".

Reply:
> But even as I type that, remember we still have a root element we can use to scope styles like we do in the stylesheet for the default UI. That's where we could use the name "asciidoc" (e.g., ".asciidoc .block").

That does not address your initial concern since it's possible to create an element within the `.asciidoc` root element via JavaScript.

For reference, Bulma is using the class `block` and does not add any class on inline elements. As soon as an inline element is within a `content` element it gets styled accordingly.

[Materialize CSS](https://materializecss.com/getting-started.html), [Foundation ](https://get.foundation/sites/docs/typography-base.html), [Bootstrap](https://getbootstrap.com/docs/5.2/content/typography/) and [Skeleton](http://getskeleton.com/#intro) are not using any class on inline elements.

Semantic UI seems to be the only one to require a `ui` class on every elements: https://semantic-ui.com/introduction/getting-started.html

[UIKit](https://getuikit.com/docs/introduction) is using the `uk` prefix:  `uk-card`, `uk-container`... 

Brad frost recommends to use a global namespace:

> All classes associated with the design system are prefixed with a global namespace, which is the Company Name followed by a hyphen
> If you’re working on a CSS architecture that is only meant to be served to a single site or if you have a lot of control over your environment, including a global namespace is likely unnecessary. But if your design system is intermingling with other technologies, it might make sense to create an identifier for system-specific code. 

https://bradfrost.com/blog/post/css-architecture-for-design-systems/

A similar approach is to use namespaces/prefixes to communicate exactly what a class does, why it exists, where (else) it might already occur, whether or not we can reuse it elsewhere, and how safe it is to bind onto or modify. 

Here's an excerpt:

> `o`-: Signify that something is an Object, and that it may be used in any number of unrelated contexts to the one you can currently see it in. Making modifications to these types of class could potentially have knock-on effects in a lot of other unrelated places. Tread carefully.
> `c-`: Signify that something is a Component. This is a concrete, implementation-specific piece of UI. All of the changes you make to its styles should be detectable in the context you’re currently looking at. Modifying these styles should be safe and have no side effects.
> (...)

https://csswizardry.com/2015/03/more-transparent-ui-code-with-namespaces/

Since it exceeds the scope of this pull request, I think we should discuss that in a dedicated issue to decide which naming convention/CSS principles we want to use. What do you think?

Reply:
follow-up issue: https://github.com/asciidoctor/asciidoctor/issues/4349

Reply:
In the AsciiDoc Language specification, we have decided to describe the main text of a list item as the principal text (or principal, for short). So I still think "principal" is the right word here. principal is defined as "the most important" and that's precisely what it is.